presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -542,15 +632,61 @@ public Duration getTotalCpuTime()
     public void start()
     {
         if (started.compareAndSet(false, true)) {
-            executor.submit(this::schedule);
+            startScheduling(getStagesReadyForExecution());
+        }
+    }
+
+    private List<SqlStageExecution> getStagesReadyForExecution()
+    {
+        return stream(forTree(StreamingPlanSection::getChildren).depthFirstPreOrder(sectionedPlan))
+                .filter(this::isReadyForExecution)
+                .flatMap(section -> stream(forTree(SubPlan::getChildren).depthFirstPreOrder(section.getPlan())))
+                .map(SubPlan::getFragment)
+                .map(PlanFragment::getId)
+                .map(this::getFragmentExecution)
+                .collect(toImmutableList());
+    }
+
+    private boolean isReadyForExecution(StreamingPlanSection section)
+    {
+        SqlStageExecution stage = getFragmentExecution(section.getPlan().getFragment().getId());
+        if (stage.getState() != PLANNED) {
+            return false;
+        }
+        for (StreamingPlanSection child : section.getChildren()) {
+            SqlStageExecution childRootStage = getFragmentExecution(child.getPlan().getFragment().getId());
+            if (childRootStage.getState() != FINISHED) {
```

Comment:
hmm.. we need to double check if it works as expected if child stage get `CANCELLED` / `ABORTED` / `FAILED` .. :)

Reply:
It can be `CANCELLED` in 2 cases

- User cancelled the query. In this case there's no need to schedule anything more.
- Parent stage finished execution (e.g.: LIMIT). In this case scheduling of the next section will be triggered by the FINISH even from the parent stage.